csrc/mma_type.h

```diff
@@ -71,26 +81,119 @@ struct MatMulTileOptions {
   }
 };
 
+enum class MmaMacro : uint64_t;
+
+struct MmaMacroUnderlying {
+  enum class Arch { NoMma, Volta, Turing, Ampere, Hopper } arch : 16;
+  unsigned m : 16;
+  unsigned n : 16;
+  unsigned k : 16;
+
+  constexpr operator uint64_t() {
+#if IS_CPP20 && !defined(__clang__)
+    // std::bit_cast for bit field is not supported by clang yet
+    return std::bit_cast<uint64_t>(*this);
+#else
+    return (uint64_t)arch << 48 | (uint64_t)m << 32 | (uint64_t)n << 16 |
+        (uint64_t)k;
+#endif
+  }
+
+  constexpr operator MmaMacro();
+
+  constexpr MmaMacroUnderlying(MmaMacro macro);
+
+  constexpr MmaMacroUnderlying(Arch arch, unsigned m, unsigned n, unsigned k)
+      : arch(arch), m(m), n(n), k(k) {}
+};
+
+static_assert(sizeof(MmaMacroUnderlying) == sizeof(uint64_t));
+
+//! Type of mma instrinsic macro to use
+//!  This will translate to which mma intrinsic from runtime string
+//!    to be generated to implement the mma op. The current plan
+//!    is to have exactly one macro for each
+//!  (arch, datatype, operand layout) triple, though there
+//!  exists multiple possibilities for some cases, e.g. for Turing and fp16
+//!  one can use 16_8_8 or 16_8_16.
+//! Will consider adding more choices that the scheduler can pick from
+//!  when our perf target becomes more fine grained, which is more likely in
+//!  latency bound kernels.
+
+#define MACRO(arch, m, n, k) \
+  arch##_##m##_##n##_##k =   \
+      MmaMacroUnderlying(MmaMacroUnderlying::Arch::arch, m, n, k)
+
+enum class MmaMacro : uint64_t {
```

Comment:
I must be missing the big picture because I haven't read your subsequent PRs. Why do we need this enum? In other words, what can this achieve while `MmaMacroUnderlying` can't?

Reply:
I think the biggest benefit is for being able to check that we are using a valid macro at compilation time. For example, if in your heuristics, you wrote something like
```C++
if (blabla) {
  macro = MmaMacro::Turing_16_7_16;
}
```
you will get a compilation error at exactly that line of code, because `MmaMacro::Turing_16_7_16` is not a valid macro. 

However, without being an enum, and if you did:
```C++
if (blabla) {
  macro = MmaMacroUnderlying(Turing, 16, 7, 16);
}
```
your code will compile correctly, and during lowering, you likely get ptxas error. The former is much easier to debug.

Reply:
I see -- the enum defines a set of **valid** <arch,m,n,k> combinations. 

Naming nit: MmaMacroUnderlying is a bit confusing because the underlying type of MmaMacro is really uint64_t. Now that I understand the enum defines a set of valid combinations, you may want to `s/MmaMacro/ValidMmaMacro` and `s/MmaMacroUnderlying/MmaMacro`. However, if you are going to use the enum form way more often, calling it `MmaMacro` makes more sense.

Reply:
I do think we will use the enum form more often. I will keep its name `MmaMacro`. And I renamed `MmaMacroUnderlying` to `MmaMacroEncode`, hope this makes it less confusing.